File: sort/insertion/Insertion.java
public class Insertion {

    public static void main(String[] args) {
        Integer[] oo = new Integer[]{34, 8, 64, 51, 32, 21};
	    sort(oo);
        for (Integer var : oo) {
            System.out.println(var);
        }
    }

    public static <T> void sort(T[] array) {
        for (int i = 1; i< array.length; i++) {
            T it = array[i];
            for (int j = i; j > 0; j--) {
                T jt = array[j - 1];
                if (compare(jt, it) < 0) {
                    break;
                }else {
                    array[j] = jt;
                    array[j - 1] = it;
                }
            }
        }
    }

    public static <T> int compare(T o1, T o2) {
        return (Integer)o1 - (Integer)o2;
    }

}

File: sort/insertion/insertion.go
package insertion

// InsertionSort 插入排序
func InsertionSort(array []int) {
	n := 0
	// 从1开始遍历,因为0前面没有数值
	for i := 1; i < len(array); i++ {
		// 在此处放置与在 代码1 处放置效果一样
		// 因为若 jv > iv 的话 ， j 与 j-1 调换位置， 下一次循环时 array[j] 依然是当初的 iv
		iv := array[i]
		for j := i; j >= 0 && array[j-1] >= iv; j-- {
			// 代码1  iv := array[j]

			// 如果 jv > iv , 调换两个元素之间的位置
			n++
			array[j], array[j-1] = array[j-1], iv

		}
	}
	println("----------------", n)
}
